### PR TITLE
Fixes #767: Second part of "no-index" changes: allow large non-indexed String values, implicit allow for "$vector"

### LINKED ISSUE #767: [Indexing options] Shredder changes to index fields
Shredder changes to use the indexing configuration to allow or deny indexing of a field

### PR DESCRIPTION
**Pull request by tatu-at-datastax**
**What this PR does**:

Completes changes needed by #767 by:

1. Adding more testing for Shredder
2. Fix the issue wrt dropping indexing of "$vector" (needs handling similar to `_id` for query projections, i.e. always index)
3. Changing validation to ignore length of non-Indexed Strings (only limited by max doc size)

These are resolved by refactoring validation logic into smaller parts, run at different times during shredding process.

**Which issue(s) this PR fixes**:
Fixes #767

**Checklist**
- [ ] Changes manually tested
- [x] Automated Tests added/updated
- [ ] Documentation added/updated
- [x] CLA Signed: [DataStax CLA](https://cla.datastax.com/)

### CODE REVIEW

File: src/main/java/io/stargate/sgv2/jsonapi/service/shredding/Shredder.java [hunk 1]
```diff
@@ -88,14 +88,21 @@ public WritableShreddedDocument shred(JsonNode doc, UUID txId, DocumentProjector
     // And then we can validate the document size
     validateDocumentSize(documentLimits, docJson);
 
-    // Before shredding, may need to drop "non-indexed" fields:
+    final WritableShreddedDocument.Builder b =
```
Comment (tatu-at-datastax) [inline src/main/java/io/stargate/sgv2/jsonapi/service/shredding/Shredder.java:91]: We need document builder immediately, to build index incrementally: in particular, `$vector` indexing should precede "no index" Projection that would typically drop it.

**Review by vkarpov15 — COMMENTED** [review]
Looks good so far :+1:

**Comment by tatu-at-datastax** [comment]
Remaining issue: when returning `after` Document for `findOneAndXxx()` need to retain it without pruning for "no-index", may need to create a deep copy.

File: src/main/java/io/stargate/sgv2/jsonapi/service/shredding/Shredder.java (hunk 1 above)
Comment (maheshrajamani) [reply src/main/java/io/stargate/sgv2/jsonapi/service/shredding/Shredder.java:93]: Any reason why it needs to be if `allow` is specified, can't we set up required indexing fields with `$vector` in it?

**Review by maheshrajamani — APPROVED** [review]
Overall looks good except for the clarification on `$vector` field

**Comment by tatu-at-datastax** [comment]
Sounds like implicit addition of `$vector` for non-empty `"allow"` -- similar to `_id` for filter Projection -- is what we'll be going with. Users can still add it to `"deny"` list if they really want to exclude it from being indexed; but that seems like an odd use case (why add `$vector` in Collection at all then?).

**Comment by tatu-at-datastax** [comment]
Need to make change for auto-inclusion of `$vector`, moving back to Draft.

File: src/main/java/io/stargate/sgv2/jsonapi/service/shredding/Shredder.java (hunk 1 above)
Comment (tatu-at-datastax) [reply src/main/java/io/stargate/sgv2/jsonapi/service/shredding/Shredder.java:93]: Will proceed with this idea, change code here.

**Comment by tatu-at-datastax** [comment]
Ready for re-review.

**Comment by tatu-at-datastax** [comment]
@maheshrajamani "$vector" handling changed as discussed: any other concerns? (I know you have approved but wanted to double-check before merging)

**Comment by maheshrajamani** [comment]
> @maheshrajamani "$vector" handling changed as discussed: any other concerns? (I know you have approved but wanted to double-check before merging)

@tatu-at-datastax Changes LGTM

**Review by vkarpov15 — COMMENTED** [review]
LGTM

**Pull request merged**